cmd/medius/images/verify.go

```diff
@@ -0,0 +1,252 @@
+				options.VerifyImagesOptions.ClusterRegistry = options.Registry
+			}
+
+			// Silence the kubevirt client log
+			kvirtlog.Log = kvirtlog.MakeLogger(kvirtlog.NullLogger{})
+			client, err := kvirtcli.GetKubevirtClient()
+			if err != nil {
+				logrus.Fatal(err)
+			}
+
+			wg, errChan := spawnWorkers(options.VerifyImagesOptions.Workers, options.Focus, func(a api.Artifact) error {
+				return verifyArtifact(a, options, client)
+			})
+			wg.Wait()
+
+			select {
+			case <-errChan:
+				os.Exit(1)
+			default:
+				os.Exit(0)
+			}
+		},
+	}
+	verifyCmd.Flags().IntVar(&options.VerifyImagesOptions.Workers, "workers", options.VerifyImagesOptions.Workers, "Number of parallel workers")
+	verifyCmd.Flags().StringVar(&options.VerifyImagesOptions.ClusterRegistry, "cluster-registry", "", "Registry to use inside the cluster")
+	verifyCmd.Flags().StringVar(&options.VerifyImagesOptions.Namespace, "namespace", options.VerifyImagesOptions.Namespace, "Namespace to run verify in")
+	verifyCmd.Flags().IntVar(&options.VerifyImagesOptions.Timeout, "timeout", options.VerifyImagesOptions.Timeout, "Maximum seconds to wait for VM to be running")
+	verifyCmd.Flags().AddGoFlagSet(kvirtcli.FlagSet())
+
+	return verifyCmd
+}
+
+func verifyArtifact(artifact api.Artifact, options *common.Options, client kvirtcli.KubevirtClient) (err error) {
+	log := common.Logger(artifact)
+
+	imgRefs, err := findImgRefs(artifact, options, log)
+	if err != nil {
+		log.WithError(err).Error("Failed to find containerdisks")
+		return err
+	}
+
+	if len(imgRefs) == 0 {
+		log.Infof("Found no containerdisks to verify")
+		return nil
+	}
+
+	vm, privateKey, err := createVM(artifact, options, imgRefs[0])
+	if err != nil {
+		log.WithError(err).Error("Failed to create VM object")
+		return err
+	}
+
+	vmClient := client.VirtualMachine(options.VerifyImagesOptions.Namespace)
+	log.Info("Creating VM")
+	if vm, err = vmClient.Create(vm); err != nil {
+		log.WithError(err).Error("Failed to create VM")
+		return err
+	}
+
+	defer func() {
+		if err = vmClient.Delete(vm.Name, &metav1.DeleteOptions{GracePeriodSeconds: pointer.Int64(0)}); err != nil {
+			log.WithError(err).Error("Failed to delete VM")
+		}
+	}()
+
+	log.Info("Waiting for VM to be ready")
+	if err = waitVMReady(vm.Name, vmClient, options.VerifyImagesOptions.Timeout); err != nil {
+		log.WithError(err).Error("VM not ready")
+		return err
+	}
+
+	vmi, err := client.VirtualMachineInstance(options.VerifyImagesOptions.Namespace).Get(vm.Name, &metav1.GetOptions{})
+	if err != nil {
+		log.WithError(err).Error("Failed to get VMI")
+		return err
+	}
+
+	log.Info("Running tests on VMI")
+	for _, testFn := range artifact.Tests() {
+		if err = testFn(vmi, &api.ArtifactTestParams{Username: VerifyUsername, PrivateKey: privateKey}); err != nil {
+			log.WithError(err).Error("Failed to verify containerdisk")
+			return err
+		}
+	}
+
+	imgRefs = append(imgRefs, prepareTimestampTag(options.Registry, artifact.Metadata()))
```

Comment:
Honestly i would just have kept the verification state in a temporary local file. All this annotation writing complicates things quite a bit. Would have made any metada change after the initial build and push to staging not necessary.
I also saw in the commit log that you moved things like e.g. timestamp tagging. Would probably also have been ok to keep as is.
Did you consider that?

Reply:
I considered using a temporary local file but I thought it would be nicer to keep all state in one place. Images can also be queried for the annotation later. However I must admit that using OCIv1 images is a bit more complex.

Moving the timestamp tagging was necessary because there is no easy way to query a registry for images with a specific label. Therefore I would have needed to iterate the repository to find all images that need to be verified. But I'm not sure if that is really a downside?

Reply:
> I considered using a temporary local file but I thought it would be nicer to keep all state in one place. Images can also be queried for the annotation later. However I must admit that using OCIv1 images is a bit more complex.

It is afaik also additional metadata which you transfer to the final location. You would probably have to remove the annotation then also when mirroring the image? Looks to me like separating the verification from the image itself make things a lot easier.

In general I am ok with pushing OCIv1 images at some point, but I would not have seen it related to this task.

> 
> Moving the timestamp tagging was necessary because there is no easy way to query a registry for images with a specific label. Therefore I would have needed to iterate the repository to find all images that need to be verified. But I'm not sure if that is really a downside?

Here it seems to be again easier to find the images to verify: Initial push pushes and writes all new images in an identifiable form and desired tags. We then simply compare the temp file with know providers and run the verifycation on them, without ever having to introspect the remote registry. After verification we update the temp file and just pass it to a promote command?

Would you mind giving it a try?

Reply:
@rmohr I changed the PR so a temporary file is used to store the verification state. Can you review again?